File: src/handlers/security.rs
use axum::{
    extract::Request,
    http::{header::HeaderMap, StatusCode, Uri},
    middleware::Next,
    response::{IntoResponse, Redirect, Response},
};

/// Middleware to enforce HTTPS in production
pub async fn https_redirect_middleware(
    uri: Uri,
    headers: HeaderMap,
    request: Request,
    next: Next,
) -> Result<Response, StatusCode> {
    // Check if we should enforce HTTPS (this would be set via config)
    let should_force_https = std::env::var("HTTPS_ENABLED")
        .unwrap_or_else(|_| "false".to_string())
        .parse()
        .unwrap_or(false);

    if should_force_https {
        // Check if request is already HTTPS
        let is_https = headers
            .get("x-forwarded-proto")
            .and_then(|v| v.to_str().ok())
            .map(|v| v == "https")
            .unwrap_or_else(|| {
                // Fallback: check the scheme
                uri.scheme_str() == Some("https")
            });

        if !is_https {
            // Construct HTTPS URL
            let host = headers
                .get("host")
                .and_then(|v| v.to_str().ok())
                .unwrap_or("localhost");
            
            let https_url = format!("https://{}{}", host, uri.path_and_query().map(|pq| pq.as_str()).unwrap_or(""));
            
            return Ok(Redirect::permanent(&https_url).into_response());
        }
    }

    Ok(next.run(request).await)
}

/// Add security headers middleware
pub async fn security_headers_middleware(
    request: Request,
    next: Next,
) -> Response {
    let mut response = next.run(request).await;
    
    let headers = response.headers_mut();
    
    // HSTS (HTTP Strict Transport Security)
    headers.insert("strict-transport-security", "max-age=31536000; includeSubDomains".parse().unwrap());
    
    // X-Content-Type-Options
    headers.insert("x-content-type-options", "nosniff".parse().unwrap());
    
    // X-Frame-Options
    headers.insert("x-frame-options", "DENY".parse().unwrap());
    
    // X-XSS-Protection
    headers.insert("x-xss-protection", "1; mode=block".parse().unwrap());
    
    // Referrer Policy
    headers.insert("referrer-policy", "strict-origin-when-cross-origin".parse().unwrap());
    
    // Content Security Policy
    headers.insert(
        "content-security-policy",
        "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://unpkg.com https://cdnjs.cloudflare.com; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; img-src 'self' data:; connect-src 'self'".parse().unwrap()
    );
    
    response
}

File: internal/middleware/security.go
package middleware

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"os"

	"github.com/gin-gonic/gin"
)

// SecurityHeaders adds enterprise security headers to all responses
// Following OWASP security best practices for web applications
func SecurityHeaders() gin.HandlerFunc {
	return func(c *gin.Context) {
		// Generate a random nonce for this request
		nonce := generateNonce()
		c.Set("CSPNonce", nonce)

		// X-Content-Type-Options: Prevents MIME type sniffing
		c.Header("X-Content-Type-Options", "nosniff")

		// X-Frame-Options: Prevents clickjacking attacks
		c.Header("X-Frame-Options", "DENY")

		// X-XSS-Protection: Enables XSS filtering (legacy browsers)
		c.Header("X-XSS-Protection", "1; mode=block")

		// Referrer-Policy: Controls referrer information
		c.Header("Referrer-Policy", "strict-origin-when-cross-origin")

		// Content-Security-Policy: Prevents various attacks
		// Using nonce-based CSP for inline scripts and styles
		cspBase := fmt.Sprintf(
			"default-src 'self'; "+
				"script-src 'self' 'nonce-%s' 'unsafe-inline' 'wasm-unsafe-eval' https://cdn.jsdelivr.net https://d2wy8f7a9ursnm.cloudfront.net; "+
				"style-src 'self' 'nonce-%s' 'unsafe-inline' https://fonts.googleapis.com; "+
				"font-src 'self' https://fonts.gstatic.com data:; "+
				"img-src 'self' data: https:; "+
				"connect-src 'self' https://api.openai.com https://notify.bugsnag.com https://sessions.bugsnag.com https://otlp.bugsnag.com; "+
				"frame-ancestors 'none'; "+
				"base-uri 'self'; "+
				"form-action 'self';",
			nonce, nonce)

		// Only add upgrade-insecure-requests in production
		if os.Getenv("SERVER_ENV") == "production" {
			cspBase += " upgrade-insecure-requests"
		}

		c.Header("Content-Security-Policy", cspBase)

		// Strict-Transport-Security: Enforces HTTPS
		// Only add if served over HTTPS
		if c.Request.TLS != nil {
			c.Header("Strict-Transport-Security", "max-age=63072000; includeSubDomains; preload")
		}

		// Permissions-Policy: Controls browser features
		c.Header("Permissions-Policy", "camera=(), microphone=(), geolocation=(), payment=()")

		// Additional security headers
		c.Header("X-Permitted-Cross-Domain-Policies", "none")

		// Only add strict CORP/COEP/COOP in production
		if os.Getenv("SERVER_ENV") == "production" {
			c.Header("Cross-Origin-Embedder-Policy", "require-corp")
			c.Header("Cross-Origin-Opener-Policy", "same-origin")
			c.Header("Cross-Origin-Resource-Policy", "same-origin")
		}

		c.Next()
	}
}

// generateNonce creates a cryptographically secure random nonce for CSP
func generateNonce() string {
	nonce := make([]byte, 16)
	if _, err := rand.Read(nonce); err != nil {
		// Fallback to a less secure but functional nonce
		return "fallback-nonce"
	}
	return base64.StdEncoding.EncodeToString(nonce)
}
